### PR TITLE
Fix #340 and DropAfterBlockBySha/NewestSha bug.

### LINKED ISSUE #340: Reorg causes orderly shutdown on btcd when addrindex=1
btcd seems to shut itself down during or shortly after a reorg when addrindex=1. It seems to corrupt the DB as simply restarting btcd just makes it shut itself right back down on the spot. Sample logs (from a btcd running on testnet3):

> 06:44:08 2015-03-04 [INF] CHAN: REORGANIZE: Block 000000000000056c231b5aa21a97b18279cb6f79788e3a143602ed5ef76497fe is causing a reorganize.
> 06:44:08 2015-03-04 [INF] CHAN: REORGANIZE: Chain forks at 00000000b5aa277d95d779ab467243e9e2ce60a939b3c36df8c5e6b53a49fb98
> 06:44:08 2015-03-04 [INF] CHAN: REORGANIZE: Old best chain head was 000000005d68052f51ea9f034ba60499d957c404e47e0c8d0662d596eac33d70
> 06:44:08 2015-03-04 [INF] CHAN: REORGANIZE: New best chain head is 000000000000056c231b5aa21a97b18279cb6f79788e3a143602ed5ef76497fe
> 06:44:28 2015-03-04 [INF] BMGR: Processed 2 blocks in the last 21.09s (15 transactions, height 325025, 2015-03-04 07:32:48 +0000 UTC)
> 06:44:28 2015-03-04 [INF] ADXR: Indexed addresses of 2 blocks in the last 21.09s (15 transactions, height 325025, 2015-03-04 07:32:48 +0000 UTC)
> 06:49:35 2015-03-04 [INF] BMGR: Processed 1 block in the last 5m6.27s (5 transactions, height 325026, 2015-03-04 07:32:48 +0000 UTC)
> 06:49:35 2015-03-04 [INF] ADXR: Indexed addresses of 1 block in the last 5m6.27s (5 transactions, height 325026, 2015-03-04 07:32:48 +0000 UTC)
> 06:50:18 2015-03-04 [INF] BMGR: Processed 1 block in the last 43.08s (1 transaction, height 325027, 2015-03-04 07:37:21 +0000 UTC)
> 06:50:18 2015-03-04 [INF] ADXR: Indexed addresses of 1 block in the last 43.08s (1 transaction, height 325027, 2015-03-04 07:37:21 +0000 UTC)
> 06:58:25 2015-03-04 [INF] BMGR: Processed 1 block in the last 8m7.23s (208 transactions, height 325028, 2015-03-04 07:42:40 +0000 UTC)
> 06:58:25 2015-03-04 [INF] ADXR: Indexed addresses of 1 block in the last 8m7.25s (208 transactions, height 325028, 2015-03-04 07:42:40 +0000 UTC)
> 07:03:35 2015-03-04 [INF] BMGR: Processed 1 block in the last 5m9.91s (94 transactions, height 325029, 2015-03-04 08:02:41 +0000 UTC)
> 07:03:35 2015-03-04 [ERR] ADXR: Couldn't get referenced txOut (303caba4550ae84ef9a8aa4e2bb872b390436d596779998a6f3bfe46cd104946:0): requested transaction does not exist
> 07:03:35 2015-03-04 [ERR] ADXR: Unable to index transactions of block requested transaction does not exist
> 07:03:35 2015-03-04 [WRN] SRVR: Server shutting down
> 07:03:35 2015-03-04 [WRN] RPCS: RPC server shutting down

and

> 21:00:48 2015-03-04 [INF] CHAN: REORGANIZE: Block 0000000000000267f64950771bb419aaf66971d54684bb0424ce43f46c5e6517 is causing a reorganize.
> 21:00:48 2015-03-04 [INF] CHAN: REORGANIZE: Chain forks at 00000000f2582a33179947d80e3138543937501f10ba3daffef2f75898118265
> 21:00:48 2015-03-04 [INF] CHAN: REORGANIZE: Old best chain head was 0000000074eb8345e94ecf80097eda83e3dc91f0f1df4a311ca97637f149cddf
> 21:00:48 2015-03-04 [INF] CHAN: REORGANIZE: New best chain head is 0000000000000267f64950771bb419aaf66971d54684bb0424ce43f46c5e6517
> 21:00:48 2015-03-04 [INF] BMGR: Processed 1 block in the last 3m24.69s (25 transactions, height 325212, 2015-03-04 21:40:16 +0000 UTC)
> 21:00:48 2015-03-04 [INF] ADXR: Indexed addresses of 1 block in the last 3m24.69s (25 transactions, height 325212, 2015-03-04 21:40:16 +0000 UTC)
> 21:10:17 2015-03-04 [INF] BMGR: Processed 1 block in the last 9m29.43s (21 transactions, height 325213, 2015-03-04 21:42:57 +0000 UTC)
> 21:10:17 2015-03-04 [INF] ADXR: Indexed addresses of 1 block in the last 9m29.43s (21 transactions, height 325213, 2015-03-04 21:42:57 +0000 UTC)
> 21:16:49 2015-03-04 [INF] BMGR: Processed 2 blocks in the last 6m31.82s (62 transactions, height 325215, 2015-03-04 22:09:05 +0000 UTC)
> 21:16:49 2015-03-04 [WRN] BCDB: unable to decode tx block 325212 0000000000000267f64950771bb419aaf66971d54684bb0424ce43f46c5e6517 txoff 1895 txlen 191
> 21:16:49 2015-03-04 [ERR] ADXR: Unable to index transactions of block: transaction e1ee7585cd9a5b19cec0e72eaebc4353bba4cd82d97bbf3fa696c6578ecee8ac not found
> 21:16:49 2015-03-04 [WRN] SRVR: Server shutting down
> 21:16:49 2015-03-04 [WRN] RPCS: RPC server shutting down
> 21:16:49 2015-03-04 [INF] RPCS: RPC server shutdown complete
> 21:16:49 2015-03-04 [INF] ADXR: Address indexer shutting down
> 21:16:49 2015-03-04 [INF] BMGR: Block manager shutting down
> 21:16:49 2015-03-04 [INF] AMGR: Address manager shutting down
> 21:16:49 2015-03-04 [INF] SRVR: Server shutdown complete
> 21:16:49 2015-03-04 [INF] BTCD: Shutdown complete

Here's a log from a restart attempt:

> 22:33:00 2015-03-04 [INF] BTCD: Version 0.10.0-beta
> 22:33:00 2015-03-04 [INF] BTCD: Loading block database from '/home/ubuntu/.btcd/data/testnet/blocks_leveldb'
> 22:33:00 2015-03-04 [INF] BTCD: Block database loaded with block height 325215
> 22:33:00 2015-03-04 [INF] BMGR: Generating initial block node index.  This may take a while...
> 22:33:01 2015-03-04 [INF] BMGR: Block index generation complete
> 22:33:01 2015-03-04 [INF] SRVR: Server listening on 0.0.0.0:18333
> 22:33:01 2015-03-04 [INF] SRVR: Server listening on [::]:18333
> 22:33:01 2015-03-04 [INF] RPCS: RPC server listening on 0.0.0.0:18334
> 22:33:01 2015-03-04 [INF] RPCS: RPC server listening on [::]:18334
> 22:33:01 2015-03-04 [INF] ADXR: Building up address index from height 325215 to 325215.
> 22:33:01 2015-03-04 [WRN] BCDB: unable to decode tx block 325212 > 0000000000000267f64950771bb419aaf66971d54684bb0424ce43f46c5e6517 txoff 1895 txlen 191
> 22:33:01 2015-03-04 [ERR] ADXR: Unable to index transactions of block: transaction e1ee7585cd9a5b19cec0e72eaebc4353bba4cd82d97bbf3fa696c6578ecee8ac not found
> 22:33:01 2015-03-04 [WRN] SRVR: Server shutting down
> 22:33:01 2015-03-04 [WRN] RPCS: RPC server shutting down
> 22:33:01 2015-03-04 [INF] RPCS: RPC server shutdown complete
> 22:33:01 2015-03-04 [INF] ADXR: Address indexer has caught up to best height, entering maintainence mode
> 22:33:01 2015-03-04 [INF] AMGR: Loaded 1801 addresses from file '/home/ubuntu/.btcd/data/testnet/peers.json'
> 22:33:01 2015-03-04 [INF] ADXR: Address indexer shutting down
> 22:33:01 2015-03-04 [INF] BMGR: Block manager shutting down
> 22:33:01 2015-03-04 [INF] AMGR: Address manager shutting down
> 22:33:01 2015-03-04 [INF] SRVR: Server shutdown complete
> 22:33:01 2015-03-04 [INF] BTCD: Shutdown complete
> 22:33:02 2015-03-04 [INF] DISC: 6 addresses found from DNS seed testnet-seed.bitcoin.petertodd.org

### PR DESCRIPTION
**Pull request by aakselrod**
- Delete spent TX in setclearSpentData when unspent by block
  disconnect on reorg.

- Test spent TX deletion when reorg causes block disconnect.

- Test for correct NewestSha results after DropAfterBlockBySha.

- Fix DropAfterBlockBySha to update info for NewestSha.

### CODE REVIEW

File: database/reorg_test.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2013-2014 Conformal Systems LLC.
```
Comment (davecgh) [inline database/reorg_test.go]: New file.  This should be 2015.

Comment (aakselrod) [inline database/reorg_test.go]: Whoops. Will fix that.

**Comment by davecgh** [comment]
@aakselrod Thanks for hunting this down and getting it fixed.  Looks good.

OK

File: database/reorg_test.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+			err = nil
+			break
+		}
+		if err != nil {
+			break
+		}
+		if rintbuf != uint32(network) {
+			break
+		}
+		err = binary.Read(dr, binary.LittleEndian, &rintbuf)
```
Comment (jrick) [inline database/reorg_test.go]: This error isn't checked

File: database/reorg_test.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		blocklen := rintbuf
+
+		rbytes := make([]byte, blocklen)
+
+		// read block
+		dr.Read(rbytes)
```
Comment (jrick) [inline database/reorg_test.go]: Missing error check

File: database/reorg_test.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+			return
+		}
+		if !maxHash.IsEqual(blkHash) || blkHeight != i {
+			t.Errorf(
+				"NewestSha returned %v (%v), expected %v (%v)",
+				blkHeight,
+				maxHash,
+				i,
+				blkHash,
+			)
```
Comment (jrick) [inline database/reorg_test.go]: Using multiple lines for arguments like this doesn't match the rest of the codebase.

File: database/reorg_test.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+// testReorganization performs reorganization tests for the passed DB type.
+// Much of the setup is copied from the blockchain package, but the test looks
+// to see if each TX in each block in the best chain can be fetched using
+// FetchTxBySha. If not, then there's a bug.
+func testReorganization(t *testing.T, dbType string) {
+	db, teardown, err := createDB(dbType, "reorganization", true)
+	if err != nil {
+		t.Errorf("Failed to create test database (%s) %v", dbType, err)
+		return
```
Comment (jrick) [inline database/reorg_test.go]: Each `Errorf` followed by a `return` can be a `Fatalf` instead.

**Comment by jrick** [comment]
Fix looks correct but I added some suggestions for the test code.

**Comment by aakselrod** [comment]
Updated test code to comply with @jrick's suggestions.

File: database/ldb/leveldb.go [hunk 6]
```diff
@@ -546,10 +550,14 @@ func (db *LevelDb) setclearSpentData(txsha *wire.ShaHash, idx uint32, set bool)
 			spentTxList[len(spentTxList)-1] = nil
 			if len(spentTxList) == 1 {
 				// write entry to delete tx from spent pool
-				// XXX
+				db.txSpentUpdateMap[*txsha] = &spentTxUpdate{delete: true}
 			} else {
-				spentTxList = spentTxList[:len(spentTxList)-1]
-				// XXX format sTxList and set update Table
+				// This code should never be hit - aakselrod
+				return fmt.Errorf(
```
Comment (davecgh) [inline database/ldb/leveldb.go]: Argument syntax formatting here doesn't match the rest of the code base.

**Comment by davecgh** [comment]
All set.  Round 2 OK.

**Comment by jrick** [comment]
ok

**Pull request merged**